### ISSUE TITLE
SecurityException trying to update a document using ES Java API 

### ISSUE THREAD

**Issue by lgambo** [floragunncom/search-guard#283]
Hi,
first of all I want to thank you for this great and useful plugin.
I'm using Java API for elasticsearch 5.1.2 with search-guard-5v5.1.2-10 Java Library in my project.
I've configured my app to work with a node certificate (in this case the name is CN=telluser1-keystore.jks) obtained by Searchguard's online tool.

Everything works fine with prepareSearch and prepareIndex functions but when I try to make a prepareUpdate i got a SecurityException with the following stacktrace:

`ElasticsearchSecurityException[unauthenticated request indices:data/write/index for user User [name=_sg_internal, roles=[]]]
        at com.floragunn.searchguard.filter.SearchGuardFilter.apply(SearchGuardFilter.java:131)
        at org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:171)
        at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:145)
        at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:87)
        at org.elasticsearch.action.update.TransportUpdateAction.shardOperation(TransportUpdateAction.java:224)
        at org.elasticsearch.action.update.TransportUpdateAction.shardOperation(TransportUpdateAction.java:172)
        at org.elasticsearch.action.update.TransportUpdateAction.shardOperation(TransportUpdateAction.java:69)
        at org.elasticsearch.action.support.single.instance.TransportInstanceSingleOperationAction$ShardTransportHandler.messageReceived(TransportInstanceSingleOperationAction.java:247)
        at org.elasticsearch.action.support.single.instance.TransportInstanceSingleOperationAction$ShardTransportHandler.messageReceived(TransportInstanceSingleOperationAction.java:243)
        at org.elasticsearch.transport.TransportRequestHandler.messageReceived(TransportRequestHandler.java:33)
        at com.floragunn.searchguard.ssl.transport.SearchGuardSSLRequestHandler.messageReceivedDecorate(SearchGuardSSLRequestHandler.java:168)
        at com.floragunn.searchguard.transport.SearchGuardRequestHandler.messageReceivedDecorate(SearchGuardRequestHandler.java:93)
        at com.floragunn.searchguard.ssl.transport.SearchGuardSSLRequestHandler.messageReceived(SearchGuardSSLRequestHandler.java:93)
        at com.floragunn.searchguard.SearchGuardPlugin$2$1.messageReceived(SearchGuardPlugin.java:266)
        at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:69)
        at org.elasticsearch.transport.TransportService$6.doRun(TransportService.java:577)
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:527)
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)`

**This problem doesn't appear when I use the sgadmin certificate (CN=sgadmin-keystore.jks).**

_Elasticsearch.yml useful part:_
searchguard.ssl.transport.keystore_filepath: CN=telluser1-keystore.jks
searchguard.ssl.transport.keystore_password: ....
searchguard.ssl.transport.truststore_filepath: truststore.jks
searchguard.ssl.transport.truststore_password: ......
searchguard.ssl.transport.enforce_hostname_verification: false
searchguard.authcz.admin_dn:
  -CN=sgadmin

_Java transport client configuration:_
`                Settings.builder()
                        .put("cluster.name", "elasticsearch")
                        .put("client.transport.sniff", false)
                        .put("client.transport.ping_timeout", 20, TimeUnit.SECONDS)
                        .put("path.home", ".")
                        .put(SSLConfigConstants.SEARCHGUARD_SSL_TRANSPORT_KEYSTORE_FILEPATH, "CN=telluser1-keystore.jks")
                        .put(SSLConfigConstants.SEARCHGUARD_SSL_TRANSPORT_TRUSTSTORE_FILEPATH, "truststore.jks")
                        .put(SSLConfigConstants.SEARCHGUARD_SSL_TRANSPORT_KEYSTORE_PASSWORD, .....)
                        .put(SSLConfigConstants.SEARCHGUARD_SSL_TRANSPORT_TRUSTSTORE_PASSWORD, .....)
                        .put(SSLConfigConstants.SEARCHGUARD_SSL_TRANSPORT_ENFORCE_HOSTNAME_VERIFICATION, false)
                        .build()`

_Searchguard configuration's values_ are all default except for sg_internal_user.yml where I added telluser1 and sg_roles_mapping.yml where I added telluser1 to sg_all_access group:
sg_all_access:
  users:
    - admin
    - adm
    - telluser1

In conclusion, my cluster has one node, server (on which ElasticSearch is installed) runs with Java 1.8.0_91-b14 and client (on which webapp runs) with java 1.8.0_71-b15.

Thank you for your attention.

**Comment by floragunncom** [comment]
Can you provide the following informations:

* Output of `keytool -list -v -keystore CN=telluser1-keystore.jks`
* The complete "client" code (where you make the index and update requests)

You may also want to read https://floragunn.com/transport-client-authentication-authorization/

**Comment by lgambo** [comment]
**Output of keytool -list -v -keystore CN=telluser1-keystore.jks:**

Tipo keystore: JKS
Provider keystore: SUN

Il keystore contiene 1 voce

Nome alias: cn=telluser1
Data di creazione: 19-gen-2017
Tipo di voce: PrivateKeyEntry
Lunghezza catena certificati: 3
Certificato[1]:
Proprietario: CN=telluser1
Autorità emittente: CN=Tell Signing CA, OU=Tell Signing CA, O=Tell
Numero di serie: 4
Valido da: Thu Jan 19 09:22:52 CET 2017 a: Sat Jan 19 09:22:52 CET 2019
Impronte digitali certificato:
	 MD5:  05:B6:1F:65:19:E5:C3:A3:82:46:BF:C2:6A:B4:B6:61
	 SHA1: B3:85:74:89:A1:7C:4F:89:D4:6F:5D:D9:D2:DB:B6:1E:11:F4:9C:BC
	 SHA256: EB:22:C1:6C:1F:4B:08:13:29:5F:AC:01:BE:8D:D0:C9:EB:D5:51:19:61:29:EE:A7:D4:EB:EA:8A:84:F3:B1:F3
	 Nome algoritmo firma: SHA256withRSA
	 Versione: 3

Estensioni:

#1: ObjectId: 2.5.29.35 Criticality=false
AuthorityKeyIdentifier [
KeyIdentifier [
0000: E9 97 B0 02 9B 3B 76 C0   CC 4D ED 04 9D DF 17 87  .....;v..M......
0010: 7D FD 4E 7B                                        ..N.
]
]

#2: ObjectId: 2.5.29.19 Criticality=false
BasicConstraints:[
  CA:false
  PathLen: undefined
]

#3: ObjectId: 2.5.29.37 Criticality=false
ExtendedKeyUsages [
  serverAuth
  clientAuth
]

#4: ObjectId: 2.5.29.15 Criticality=true
KeyUsage [
  DigitalSignature
  Key_Encipherment
]

#5: ObjectId: 2.5.29.17 Criticality=false
SubjectAlternativeName [
  DNSName: telluser1
  OIDName: 1.2.3.4.5.5
]

#6: ObjectId: 2.5.29.14 Criticality=false
SubjectKeyIdentifier [
KeyIdentifier [
0000: 7F E0 68 47 B4 28 CA 30   B6 EC B6 91 83 35 56 61  ..hG.(.0.....5Va
0010: 08 00 A1 15                                        ....
]
]

Certificato[2]:
Proprietario: CN=Tell Signing CA, OU=Tell Signing CA, O=Tell
Autorità emittente: CN=Tell Root CA, OU=Tell Root CA, O=Tell
Numero di serie: 2
Valido da: Thu Jan 19 09:22:44 CET 2017 a: Tue Jan 19 09:22:44 CET 2027
Impronte digitali certificato:
	 MD5:  6B:36:AC:A3:18:8F:59:E4:F0:0C:21:47:64:94:C0:3D
	 SHA1: 67:EC:2D:56:B7:A2:01:5B:B3:60:A1:3D:3A:1A:D3:3A:B6:84:59:95
	 SHA256: D7:51:AB:B1:4B:F7:F3:5A:85:95:34:82:E1:F4:9B:D9:C6:14:17:D5:7E:A8:E2:68:9E:2B:72:FB:30:8F:37:40
	 Nome algoritmo firma: SHA256withRSA
	 Versione: 3

Estensioni:

#1: ObjectId: 2.5.29.35 Criticality=false
AuthorityKeyIdentifier [
KeyIdentifier [
0000: F7 E6 AF 97 B8 25 E2 EB   CE 67 EE D1 F0 6C C3 12  .....%...g...l..
0010: E8 64 4D C7                                        .dM.
]
]

#2: ObjectId: 2.5.29.19 Criticality=true
BasicConstraints:[
  CA:true
  PathLen:0
]

#3: ObjectId: 2.5.29.15 Criticality=true
KeyUsage [
  Key_CertSign
  Crl_Sign
]

#4: ObjectId: 2.5.29.14 Criticality=false
SubjectKeyIdentifier [
KeyIdentifier [
0000: E9 97 B0 02 9B 3B 76 C0   CC 4D ED 04 9D DF 17 87  .....;v..M......
0010: 7D FD 4E 7B                                        ..N.
]
]

Certificato[3]:
Proprietario: CN=Tell Root CA, OU=Tell Root CA, O=Tell
Autorità emittente: CN=Tell Root CA, OU=Tell Root CA, O=Tell
Numero di serie: 1
Valido da: Thu Jan 19 09:22:44 CET 2017 a: Tue Jan 19 09:22:44 CET 2027
Impronte digitali certificato:
	 MD5:  C8:BE:0C:73:AE:33:2D:BD:41:F4:5E:A5:14:25:3A:71
	 SHA1: B1:04:32:69:2A:94:51:A5:D8:77:34:BC:EA:39:E7:64:13:6C:7D:0E
	 SHA256: 7A:78:F2:68:42:DC:29:9F:1D:3C:EE:5C:D3:DF:01:06:0C:6D:9B:1A:C5:53:C1:5E:FB:74:4F:73:9F:F4:52:FA
	 Nome algoritmo firma: SHA256withRSA
	 Versione: 3

Estensioni:

#1: ObjectId: 2.5.29.35 Criticality=false
AuthorityKeyIdentifier [
KeyIdentifier [
0000: F7 E6 AF 97 B8 25 E2 EB   CE 67 EE D1 F0 6C C3 12  .....%...g...l..
0010: E8 64 4D C7                                        .dM.
]
]

#2: ObjectId: 2.5.29.19 Criticality=true
BasicConstraints:[
  CA:true
  PathLen:2147483647
]

#3: ObjectId: 2.5.29.15 Criticality=true
KeyUsage [
  Key_CertSign
  Crl_Sign
]

#4: ObjectId: 2.5.29.14 Criticality=false
SubjectKeyIdentifier [
KeyIdentifier [
0000: F7 E6 AF 97 B8 25 E2 EB   CE 67 EE D1 F0 6C C3 12  .....%...g...l..
0010: E8 64 4D C7                                        .dM.
]
]



*******************************************
*******************************************





**The complete "client" code (where you make the index and update requests)**

        UpdateResponse updateResponse = transportClient.prepareUpdate(".kibana","dashboard", id).setDoc(params).execute().actionGet();

SearchResponse response = transportClient.prepareSearch(".kibana").setTypes("dashboard")
                .setSearchType(SearchType.DEFAULT)
                .setQuery(QueryBuilders.simpleQueryStringQuery(user + "*")).setExplain(false)
                .setSize(100)
                .execute().actionGet();

IndexResponse indexResponse = transportClient.prepareIndex("log", "action").setSource(
                    XContentFactory.jsonBuilder()
                            .startObject()
                            .field("time", new Date())
                            .field("node", nodeid)
                            .field("action", action)
                            .endObject()).execute().actionGet();

The problem is on the first statement (and doesn't work even if I make a prepareUpdate on any other index), the other two statements works perfectly.
That issue doesn't appear if I use "CN=sgadmin-keystore.jks" certificate.


Thank you for your kind suggestions

**Comment by floragunncom** [comment]
Please use a non node-certificate (that means: use a client or user certificate) for the transport client. Then add credentials via an header as described here: https://floragunn.com/transport-client-authentication-authorization/

**Comment by floragunncom** [comment]
Does it now work for you?

**Comment by lgambo** [comment]
Sorry for the late answer, I tried to configure a client certificate (CN=demouser-keystore.jks) adding that to elasticsearch.yml (in the "searchguard.authcz.admin_dn" parameter). 
It works perfectly! Thank you
But just curious, node certificates didn't work because Elasticsearch Java API works with transport layer so it isn't a real cluster node, is that right?

**Comment by floragunncom** [comment]
You should NOT add the dn to "searchguard.authcz.admin_dn". If you do this your client can also change the Search Guard configuration. Instead you may want to use basic authentication or the impersonation feature described in https://floragunn.com/transport-client-authentication-authorization/. See also 

* https://github.com/floragunncom/search-guard/issues/264 and 
* https://github.com/floragunncom/search-guard/issues/207 and 
* https://github.com/floragunncom/search-guard/issues/53#issuecomment-217019746

**Comment by floragunncom** [comment]
No feedback, assume issue is solved
